File: validation/mml_free_sd_cross_language_probe.R
#!/usr/bin/env Rscript
# Bounded development check: 24 Persons, 2 Raters/Tasks/Criteria, scores 0:3.
# Reuse the independent R likelihood, leaving its historical runner unchanged.
args <- commandArgs(TRUE)
stopifnot(length(args) == 3L, !file.exists(args[3]))
input <- jsonlite::read_json(args[1], simplifyVector = TRUE)
stopifnot(identical(input$classification, "OBSERVED_DEVELOPMENT_ONLY"))
helpers <- c("marginal_components", "objective", "finite_difference_gradient")
for (e in parse(args[2])) {
  if (is.call(e) && identical(e[[1]], as.name("<-")) &&
      as.character(e[[2]])[1] %in% helpers) eval(e)
}
stopifnot(all(vapply(helpers, exists, logical(1), envir = .GlobalEnv, inherits = FALSE)))
data <- input$data
stopifnot(nrow(data) == 192L, all(data$Score %in% 0:3),
          identical(sort(unique(data$Person)), sprintf("P%03d", 0:23)),
          all(data$Rater %in% c("R1", "R2")),
          all(data$Task %in% c("T1", "T2")),
          all(data$Criterion %in% c("C1", "C2")))
prepared <- list(person = match(data$Person, sort(unique(data$Person))),
                 rater = match(data$Rater, c("R1", "R2")),
                 task = match(data$Task, c("T1", "T2")),
                 criterion = match(data$Criterion, c("C1", "C2")),
                 score = data$Score, n_person = 24L)
stopifnot(!is.unsorted(prepared$person), all(tabulate(prepared$person) == 8L))

# Existing statmod uses a different eigensolver and retains positive tail weights.
rules <- lapply(c(31L, 61L, 121L, 181L), statmod::gauss.quad.prob, dist = "normal")
names(rules) <- c("31", "61", "121", "181")
stopifnot(all(vapply(rules, function(x) {
  all(is.finite(x$nodes)) && all(is.finite(x$weights)) &&
    all(x$weights > 0) && abs(sum(x$weights) - 1) < 1e-12
}, logical(1))))
normal_quadrature <- function(n, sigma) {
  stopifnot(as.character(n) %in% names(rules), is.finite(sigma), sigma > 0)
  x <- rules[[as.character(n)]]
  list(nodes = x$nodes * sigma, weights = x$weights)
}
expand_parameters <- function(par) {
  stopifnot(model %in% c("RSM", "PCM"), all(is.finite(par)),
            length(par) == if (model == "RSM") 7L else 9L)
  steps <- if (model == "RSM") {
    matrix(rep(c(par[5:6], -sum(par[5:6])), 2L), nrow = 2L, byrow = TRUE)
  } else rbind(c(par[5:6], -sum(par[5:6])), c(par[7:8], -sum(par[7:8])))
  list(Rater = c(par[1], -par[1]), Task = c(par[2], -par[2]),
       Criterion = par[3:4], steps = steps, sigma = exp(tail(par, 1L)))
}
probabilities <- function(par, theta) {
  x <- expand_parameters(par)
  base <- -x$Rater[prepared$rater] - x$Task[prepared$task] - x$Criterion[prepared$criterion]
  steps <- t(apply(x$steps, 1L, function(v) c(0, cumsum(v))))
  logits <- outer(theta + base, 0:3) - steps[prepared$criterion, , drop = FALSE]
  shifted <- logits - apply(logits, 1L, max)
  exp(shifted) / rowSums(exp(shifted))
}
score_summary <- function(par, q) {
  c <- marginal_components(par, prepared, q, retain_posterior = TRUE)
  eap <- as.numeric(c$posterior %*% c$nodes)
  centered <- matrix(c$nodes, 24L, q, byrow = TRUE) - eap
  list(nll = -c$loglik, eap = eap,
       sd = sqrt(rowSums(c$posterior * centered^2)))
}

# Independent adaptive integral in standard-normal z, split at the pattern mode.
# The finite [-12,12] interval has explicit absolute normal-tail moment bounds.
continuous_summary <- function(par) {
  sigma <- expand_parameters(par)$sigma
  per_person <- lapply(seq_len(24L), function(p) {
    rows <- which(prepared$person == p)
    log_integrand <- function(z) vapply(z, function(v) {
      probs <- probabilities(par, sigma * v)
      sum(log(probs[cbind(rows, prepared$score[rows] + 1L)])) + dnorm(v, log = TRUE)
    }, numeric(1))
    mode <- optimize(log_integrand, c(-12, 12), maximum = TRUE, tol = 1e-10)
    stopifnot(abs(mode$maximum) < 11.9)
    integral <- lapply(0:2, function(k) {
      f <- function(z) exp(log_integrand(z) - mode$objective) * z^k
      parts <- lapply(list(c(-12, mode$maximum), c(mode$maximum, 12)), function(b) {
        integrate(f, b[1], b[2], subdivisions = 300L, rel.tol = 1e-10, abs.tol = 1e-12)
      })
      c(value = sum(vapply(parts, `[[`, numeric(1), "value")),
        error = sum(vapply(parts, `[[`, numeric(1), "abs.error")))
    })
    mass <- integral[[1]][["value"]]
    mean <- sigma * integral[[2]][["value"]] / mass
    variance <- sigma^2 * integral[[3]][["value"]] / mass - mean^2
    stopifnot(mass > 0, variance > 0)
    list(log_marginal = mode$objective + log(mass), eap = mean, sd = sqrt(variance),
         mode_z = mode$maximum,
         scaled_integrals = vapply(integral, `[[`, numeric(1), "value"),
         scaled_numeric_errors = vapply(integral, `[[`, numeric(1), "error"),
         log_scale = mode$objective)
  })
  list(nll = -sum(vapply(per_person, `[[`, numeric(1), "log_marginal")),
       eap = vapply(per_person, `[[`, numeric(1), "eap"),
       sd = vapply(per_person, `[[`, numeric(1), "sd"), per_person = per_person,
       absolute_tail_bounds = c(mass = 2 * pnorm(-12),
         first_absolute_theta_moment = sigma * 2 * dnorm(12),
         second_theta_moment = sigma^2 * 2 * (12 * dnorm(12) + pnorm(-12))))
}
evaluate <- function(par, q) {
  c(score_summary(par, q), list(
    gradient_h = finite_difference_gradient(par, prepared, q, step = 1e-5),
    gradient_half = finite_difference_gradient(par, prepared, q, step = 5e-6)))
}
# Analytic limiting control: zero structural effects and nearly zero latent SD
# give probability 1/4 per response and NLL 192*log(4).
for (model in c("RSM", "PCM")) {
  point <- c(rep(0, if (model == "RSM") 6L else 8L), log(1e-10))
  stopifnot(abs(objective(point, prepared, 31L) - 192 * log(4)) < 1e-10,
            max(abs(probabilities(point, 0) - 0.25)) < 1e-14)
}
cases <- lapply(seq_len(nrow(input$cases)), function(i) {
  case <- input$cases[i, ]
  model <<- case$model
  par <- as.numeric(case$coordinates[[1]])
  c(list(id = case$id, model = model, q = case$q), evaluate(par, case$q),
    list(probabilities = if (case$q == 31L) lapply(c(-3, 0, 3),
      function(theta) probabilities(par, theta)[seq_len(8L), ]) else NULL))
})
refits <- lapply(seq_len(nrow(input$fits)), function(i) {
  case <- input$fits[i, ]
  model <<- case$model
  start <- as.numeric(case$start[[1]])
  fit <- optim(start, objective, gr = finite_difference_gradient,
               prepared = prepared, q = case$q, method = "L-BFGS-B",
               lower = c(rep(-Inf, length(start) - 1L), log(0.05)),
               upper = c(rep(Inf, length(start) - 1L), log(10)),
               control = list(maxit = 250L, factr = 1, pgtol = 1e-8))
  cat(model, "Q", case$q, "R refit:", fit$convergence, "\n")
  # Post-result diagnostic after the initial development run had two code-52
  # line-search terminations. Apply once to ALL eight cases; retain primary codes.
  restart <- optim(fit$par, objective, gr = finite_difference_gradient,
                   prepared = prepared, q = case$q, method = "BFGS",
                   control = list(maxit = 100L, reltol = 1e-13))
  # Continuous comparison holds the Python point fixed, separately from this refit.
  python_par <- as.numeric(case$python_coordinates[[1]])
  list(model = model, q = case$q, coordinates = fit$par, start = start,
       convergence = fit$convergence, message = fit$message, counts = fit$counts,
       at_r_fit = evaluate(fit$par, case$q),
       diagnostic_restart = list(convergence = restart$convergence,
         coordinates = restart$par, counts = restart$counts,
         displacement = max(abs(restart$par - fit$par)),
         nll_gain = fit$value - restart$value, at_point = evaluate(restart$par, case$q)),
       continuous_at_python_fit = continuous_summary(python_par))
})
jsonlite::write_json(list(classification = "OBSERVED_DEVELOPMENT_ONLY",
  scientific_inference_ready = FALSE, rules = rules, cases = cases, refits = refits,
  runtime = list(R = R.version.string, platform = R.version$platform,
    statmod = as.character(packageVersion("statmod")),
    jsonlite = as.character(packageVersion("jsonlite")), LAPACK = La_version(),
    BLAS = unname(extSoftVersion()["BLAS"]))), args[3], digits = NA,
  pretty = TRUE, auto_unbox = TRUE, na = "null", null = "null")

File: validation/mml_free_sd_cross_language_probe.py
#!/usr/bin/env python3
"""Cross-check the observed 24-Person RSM/PCM fixture; no inference promotion.

Reuses retained Q31/Q61 points, the app's v2 kernel, and the independent R
likelihood. Additional Q121/Q181 fits and continuous evaluations are development
diagnostics. Requires R/statmod/jsonlite and the existing Python test environment.
"""
from dataclasses import replace
import argparse
import hashlib
import json
from pathlib import Path
import platform
import runpy
import subprocess
import sys

import numpy as np
import scipy

ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))
import streamlit_app as app
from mfrm_app.mml_engine_v2 import run_free_sd_stationarity_v2
from mfrm_app.mml_stationarity import JointPolishOptions, make_joint_free_sd_functions
from validation.mml_free_sd_stationarity_adapter import prepare_app_free_sd_problem

CLASSIFICATION = "OBSERVED_DEVELOPMENT_ONLY"
ORDERS = (31, 61, 121, 181)
# Implementation checks fixed before this probe's results; not scientific margins.
LIMITS = {"nll": 1e-9, "gradient": 1e-6, "probability": 1e-12,
          "eap": 1e-10, "sd": 1e-10, "node": 1e-12,
          "relative_weight": 1e-10, "moment": 1e-10,
          "refit_coordinate": 1e-4, "refit_nll": 1e-8, "refit_score": 1e-4}


def dump(path, value):
    with path.open("x") as handle:
        json.dump(value, handle, indent=2, allow_nan=False)
        handle.write("\n")


def difference(left, right):
    a, b = np.asarray(left, dtype=float), np.asarray(right, dtype=float)
    if a.shape != b.shape or not np.isfinite(a).all() or not np.isfinite(b).all():
        raise ValueError("Nonfinite or misaligned cross-language evidence")
    return float(np.max(np.abs(a - b)))


def evaluate_python(problem, coordinates):
    point = np.asarray(coordinates, dtype=float)
    _, value_gradient = make_joint_free_sd_functions(problem.value, problem.value_gradient)
    value, gradient = value_gradient(point)
    params = app.expand_params(point[:-1], problem.sizes, problem.config)
    quad = app.gauss_hermite_normal(problem.quadrature_points, sd=np.exp(point[-1]))
    scores = app.compute_person_eap(problem.idx, problem.config, params, quad)
    base = app.compute_base_eta(problem.idx, params, problem.config)
    if problem.config["model"] == "RSM":
        cumulative = np.r_[0, np.cumsum(params["steps"])]
        probabilities = [app.category_prob_rsm(base + t, cumulative)[:8] for t in (-3, 0, 3)]
    else:
        cumulative = np.column_stack([np.zeros(2), np.cumsum(params["steps_mat"], axis=1)])
        probabilities = [app.category_prob_pcm(base + t, cumulative, problem.idx["step_idx"])[:8]
                         for t in (-3, 0, 3)]
    return {"nll": float(value), "gradient": gradient.tolist(),
            "eap": scores.Estimate.tolist(), "sd": scores.SD.tolist(),
            "probabilities": np.asarray(probabilities).tolist()}


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--output-dir", type=Path, required=True)
    args = parser.parse_args()
    output = args.output_dir.resolve()
    output.mkdir(parents=True, exist_ok=False)
    retained_path = ROOT / "validation/mml_free_sd_development_probe_20260911.json"
    retained = json.loads(retained_path.read_text())
    hashes = dict(retained["source_sha256"])
    for name, expected in hashes.items():
        if hashlib.sha256((ROOT / name).read_bytes()).hexdigest() != expected:
            raise ValueError(f"Retained point source changed: {name}")
    for path in (Path(__file__), Path(__file__).with_suffix(".R"), retained_path,
                 ROOT / "validation/known_assignment_mml_crossfit.R"):
        hashes[str(path.relative_to(ROOT))] = hashlib.sha256(path.read_bytes()).hexdigest()
    test = runpy.run_path(str(ROOT / "tests/test_mml_free_sd_quadrature_adapter.py"))
    source = test["native_result"].__wrapped__()
    data = source["prep"]["data"][["Person", "Rater", "Task", "Criterion", "Score"]].copy()
    assert len(data) == 192 and data.Person.nunique() == 24
    assert data.Person.is_monotonic_increasing
    assert len(data.iloc[:8].drop_duplicates(["Rater", "Task", "Criterion"])) == 8
    problems, python_cases, python_fits, cases, fits = {}, {}, {}, [], []
    for record in retained["records"]:
        model = record["model"]
        native = source if model == "RSM" else app.mfrm_estimate(
            data, person_col="Person", facet_cols=["Rater", "Task", "Criterion"],
            score_col="Score", rating_min=0, rating_max=3, model="PCM", method="MML",
            noncenter_facet="Criterion", step_facet="Criterion", mml_engine="EM",
            quad_points=9, estimate_population_sd=True, population_prior_sd=1.0,
            maxit=300, reltol=1e-6, min_obs_per_element=1, min_obs_per_category=1)
        problem = prepare_app_free_sd_problem(native, quadrature_points=31)
        assert list(problem.sizes.values()) == ([0, 0, 1, 1, 2, 2] if model == "RSM"
                                                else [0, 0, 1, 1, 2, 4])
        assert problem.config["facet_names"] == ["Rater", "Task", "Criterion"]
        assert all(problem.config["facet_signs"][f] == -1 for f in problem.config["facet_names"])
        assert np.all(problem.idx["weight"] == 1)
        for f, levels in (("Rater", ["R1", "R2"]), ("Task", ["T1", "T2"]),
                          ("Criterion", ["C1", "C2"])):
            assert list(problem.config["facet_levels"][f]) == levels
        old_runs = {31: record["raw"]["primary_run"], 61: record["raw"]["sensitivity_run"]}
        legacy = np.asarray(old_runs[31]["primary_polish"]["initial_joint_coordinates"])
        points = {"legacy": legacy}
        points.update({f"q{q}": np.asarray(run["restart_polish"]["joint_coordinates"])
                       for q, run in old_runs.items()})
        points["perturbed"] = points["q61"] + np.linspace(-0.15, 0.15, len(legacy))
        for q in ORDERS:
            current = replace(problem, quadrature_points=q)
            problems[model, q] = current
            for label, point in points.items():
                key = f"{model}:{label}:Q{q}"
                cases.append({"id": key, "model": model, "q": q, "coordinates": point.tolist()})
                python_cases[key] = evaluate_python(current, point)
            if q in old_runs:
                run = old_runs[q]
            else:
                start = points["q61"]
                run = run_free_sd_stationarity_v2(start[:-1], np.exp(start[-1]),
                    current.value, current.value_gradient, observations=current.observations,
                    structural_bounds=current.structural_bounds, sigma_bounds=current.sigma_bounds,
                    options=JointPolishOptions(maxiter=250, gtol=1e-8, ftol=1e-15),
                    constraint_residual_function=current.constraint_residual).to_dict()
            point = run["restart_polish"]["joint_coordinates"]
            python_fits[f"{model}:Q{q}"] = {"run": run, "values": evaluate_python(current, point)}
            fits.append({"model": model, "q": q, "start": legacy.tolist(), "python_coordinates": point})
            print(f"{model} Q{q}: Python point ready", flush=True)
    dump(output / "input.json", {"classification": CLASSIFICATION, "limits": LIMITS, "source_sha256": hashes,
        "data": data.to_dict(orient="records"), "cases": cases, "fits": fits})
    dump(output / "python.json", {"cases": python_cases, "fits": python_fits})
    subprocess.run(["Rscript", "--vanilla", str(Path(__file__).with_suffix(".R")),
        str(output / "input.json"), str(ROOT / "validation/known_assignment_mml_crossfit.R"),
        str(output / "r.json")], check=True, cwd=ROOT)
    r = json.loads((output / "r.json").read_text())
    assert r["classification"] == CLASSIFICATION and r["scientific_inference_ready"] is False
    assert [row["id"] for row in r["cases"]] == [row["id"] for row in cases]
    assert [(row["model"], row["q"]) for row in r["refits"]] == list(problems)
    rule_checks, cross_checks, refit_checks, movements = [], [], [], []
    for q in ORDERS:
        rq = r["rules"][str(q)]
        py = app.gauss_hermite_normal(q)
        w, x = np.asarray(rq["weights"]), np.asarray(rq["nodes"])
        assert np.all(w > 0)
        check = {"q": q, "node": difference(x, py["nodes"]),
                 "relative_weight": float(np.max(np.abs(w / py["weights"] - 1))),
                 "moment": difference([w.sum(), w @ x, w @ x**2, w @ x**4], [1, 0, 1, 3])}
        check["pass"] = all(check[k] <= LIMITS[k] for k in ("node", "relative_weight", "moment"))
        rule_checks.append(check)
    for row in r["cases"]:
        py = python_cases[row["id"]]
        check = {"id": row["id"], "nll": abs(py["nll"] - row["nll"]),
                 "gradient": max(difference(py["gradient"], row["gradient_h"]),
                                 difference(py["gradient"], row["gradient_half"])),
                 "eap": difference(py["eap"], row["eap"]), "sd": difference(py["sd"], row["sd"]),
                 "probability": difference(py["probabilities"], row["probabilities"])
                                if row["probabilities"] is not None else 0.0}
        check["pass"] = all(check[k] <= LIMITS[k] for k in ("nll", "gradient", "eap", "sd", "probability"))
        cross_checks.append(check)
    for row in r["refits"]:
        model, q = row["model"], row["q"]
        py = python_fits[f"{model}:Q{q}"]
        at_r = evaluate_python(problems[model, q], row["coordinates"])
        check = {"model": model, "q": q, "r_convergence": row["convergence"],
                 "refit_coordinate": difference(row["coordinates"], py["run"]["restart_polish"]["joint_coordinates"]),
                 "refit_nll": abs(at_r["nll"] - py["values"]["nll"]),
                 "refit_score": max(abs(np.asarray(at_r["gradient"]))),
                 "r_point_nll_parity": abs(at_r["nll"] - row["at_r_fit"]["nll"]),
                 "r_point_gradient_parity": difference(at_r["gradient"], row["at_r_fit"]["gradient_half"])}
        check["pass"] = bool(row["convergence"] == 0 and all(check[k] <= LIMITS[k] for k in
            ("refit_coordinate", "refit_nll", "refit_score")) and check["r_point_nll_parity"] <= LIMITS["nll"]
            and check["r_point_gradient_parity"] <= LIMITS["gradient"])
        check["numeric_checks_pass"] = bool(all(check[k] <= LIMITS[k] for k in
            ("refit_coordinate", "refit_nll", "refit_score")) and check["r_point_nll_parity"] <= LIMITS["nll"]
            and check["r_point_gradient_parity"] <= LIMITS["gradient"])
        restart = row["diagnostic_restart"]
        restart_py = evaluate_python(problems[model, q], restart["coordinates"])
        check["diagnostic_restart"] = {"convergence": restart["convergence"],
            "displacement": restart["displacement"], "nll_gain": restart["nll_gain"],
            "python_score": float(np.max(np.abs(restart_py["gradient"]))),
            "nll_parity": abs(restart_py["nll"] - restart["at_point"]["nll"]),
            "gradient_parity": difference(restart_py["gradient"], restart["at_point"]["gradient_half"])}
        refit_checks.append(check)
        continuous = row["continuous_at_python_fit"]
        movements.append({"model": model, "q": q,
            "sigma": py["run"]["final_sigma"], "joint_score": py["run"]["final_projected_gradient_supnorm"],
            "nll": py["values"]["nll"], "continuous_nll_at_same_point": continuous["nll"],
            "nll_vs_continuous": abs(py["values"]["nll"] - continuous["nll"]),
            "eap_vs_continuous": difference(py["values"]["eap"], continuous["eap"]),
            "sd_vs_continuous": difference(py["values"]["sd"], continuous["sd"])})
    for name, expected in hashes.items():
        assert hashlib.sha256((ROOT / name).read_bytes()).hexdigest() == expected, name
    result = {"classification": CLASSIFICATION, "qualification_eligible": False,
        "scientific_inference_ready": False, "limits": LIMITS, "source_sha256": hashes,
        "runtime": {"python": platform.python_version(), "numpy": np.__version__, "scipy": scipy.__version__, **r["runtime"]},
        "fixture": {"generation_model": "RSM", "seed": 20260811, "persons": 24, "ratings": 192},
        "quadrature": rule_checks, "cross_evaluation": cross_checks, "refits": refit_checks,
        "integration_at_python_fits": movements,
        "artifact_sha256": {p.name: hashlib.sha256(p.read_bytes()).hexdigest() for p in output.iterdir()}}
    result["implementation_checks_pass"] = all(c["pass"] for c in rule_checks + cross_checks + refit_checks)
    result["same_point_checks_pass"] = all(c["pass"] for c in rule_checks + cross_checks)
    result["refit_numeric_checks_pass"] = all(c["numeric_checks_pass"] for c in refit_checks)
    dump(output / "summary.json", result)
    print(json.dumps({"implementation_checks_pass": result["implementation_checks_pass"],
                      "integration_at_python_fits": movements}, indent=2), flush=True)
    if not result["implementation_checks_pass"]:
        raise SystemExit("Development comparison needs review; all raw results retained.")


if __name__ == "__main__":
    main()
